varnish-sys/src/vcl/ws.rs

```diff
@@ -41,105 +45,129 @@ impl<'a> Workspace<'a> {
+    ///
+    /// Will panic if `size` is 0.
+    pub fn alloc_buffer(&mut self, size: usize) -> Result<&'a mut [MaybeUninit<u8>], VclError> {
+        let ptr = unsafe { self.alloc(size) };
+        if ptr.is_null() {
+            Err(VclError::WsOutOfMemory(size))
+        } else {
+            Ok(unsafe { from_raw_parts_mut(ptr.cast(), size) })
         }
     }
 
-    /// Copy any `AsRef<[u8]>` into the workspace
-    pub fn copy_bytes<T: AsRef<[u8]>>(&mut self, src: &T) -> Result<&'a [u8], VclError> {
-        let buf = src.as_ref();
-        let l = buf.len();
+    /// Allocate memory on Workspace, and move a value into it.
+    /// The value will be dropped in case of out of memory error.
+    pub fn alloc_value<T>(&mut self, value: T) -> Result<&'a mut T, VclError> {
+        let size = size_of::<T>();
+        assert!(size > 0, "Type {} has sizeof=0", type_name::<T>());
 
-        let dest = unsafe { self.alloc(l)? };
-        dest.copy_from_slice(buf);
-        Ok(dest)
+        let val = unsafe { self.alloc(size).cast::<T>().as_mut() };
+        let val = val.ok_or(VclError::WsOutOfMemory(size_of::<T>()))?;
+        *val = value;
+        Ok(val)
     }
 
-    /// Copy any `AsRef<CStr>` into the workspace
-    pub fn copy_cstr<T: AsRef<CStr>>(&mut self, src: &T) -> Result<VCL_STRING, VclError> {
-        let buf = self.copy_bytes(&src.as_ref().to_bytes_with_nul())?;
-        Ok(VCL_STRING(buf.as_ptr().cast::<c_char>()))
+    /// Copy any `AsRef<[u8]>` into the workspace
+    fn copy_bytes(&mut self, src: impl AsRef<[u8]>) -> Result<&'a [u8], VclError> {
+        // Re-implement unstable `maybe_uninit_write_slice` and `maybe_uninit_slice`
+        // See https://github.com/rust-lang/rust/issues/79995
+        // See https://github.com/rust-lang/rust/issues/63569
+        let src = src.as_ref();
+        let dest = self.alloc_buffer(src.len())?;
+        dest.copy_from_slice(maybe_uninit(src));
+        Ok(slice_assume_init_mut(dest))
     }
 
-    /// Same as [`Workspace::copy_bytes`] but adds NULL character at the end to help converts buffers into
-    /// `VCL_STRING`s. Returns an error if `src` contain NULL characters.
-    pub fn copy_bytes_with_null<T: AsRef<[u8]>>(&mut self, src: &T) -> Result<&'a CStr, VclError> {
-        let buf = src.as_ref();
-        if buf.contains(&0) {
-            return Err("source buffer contains NULL character".into());
-        }
-        let l = buf.len();
-        let dest = unsafe { self.alloc(l + 1)? };
-        dest[..l].copy_from_slice(buf);
-        dest[l] = b'\0';
-        // Safe because there are no NULLs in the source, and we just added a NULL
-        Ok(unsafe { CStr::from_bytes_with_nul_unchecked(dest) })
+    /// Copy any `AsRef<[u8]>` into a new [`VCL_BLOB`] stored in the workspace
+    pub fn copy_blob(&mut self, value: impl AsRef<[u8]>) -> Result<VCL_BLOB, VclError> {
+        let buf = self.copy_bytes(value)?;
+        let blob = self.alloc_value(vrt_blob {
```

Comment:
Performance-wise, we don't care, but it feels a bit silly to create a struct on the stack to copy it, would we be better creating it in the workspace then populating it directly?

Reply:
how do you see it?  Let's say we have a slightly different API:

```rust
// Since we don't have any way of dropping T,
// we should use MaybeUninit<T> at first,
// and later switch to ManualDrop<T>...
// not that it matters because we never returned owned objects, only &mut T
fn alloc_type<T>(&mut self) -> VclResult<&mut MaybeUninit<T>>;

// usage
let val = ws.alloc_type::<vrt_blob>();
val.write( vrt_blob { ... } );  // this results in pretty much the same as what I have
```

Alternatively, we can have uninitiized API:

```rust
// result is UB because T is kinda "owned" by self, but hasn't been initialized yet
unsafe fn alloc_type<T>(&mut self) -> VclResult<&mut T>;

// usage
let val = ws.alloc_type::<vrt_blob>();
val.blob = ptr::from_ref(buf).cast::<c_void>();
val.len = buf.len();
val.type_ = Default::default();
```

In short, given that we only deal with a 3 field (24 byte) value, you either copy these 24 bytes 8 byte at a time and eat a cost of dereferencing on each assignment, or copy all 24 bytes at once in one operation...  And now we are deep in the realm of compiler optimization - how smart will LLVM+Rust be to make this much simpler...  I doubt it will play a role.

Reply:
fair, let's optimize for human readability. Ignore my comment!